File: April 2020/29 - Binary Tree Maximum Path Sum/Solution.java
class Solution {
    int max_sum = Integer.MIN_VALUE;
    public int maxPathSum(TreeNode root) {
        max_gain(root);
        return max_sum;
    }
    
    int max_gain(TreeNode node) {
        if(node == null)
            return 0;
        
        int leftGain = Math.max(max_gain(node.left), 0);
        int rightGain = Math.max(max_gain(node.right), 0);
        max_sum = Math.max(max_sum, node.val + leftGain + rightGain);
        
        return node.val + Math.max(leftGain, rightGain);
    }
}

File: April 2020/29 - Binary Tree Maximum Path Sum/Solution.py
class Solution(object):
    def maxPathSum(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        self.max = float('-inf')
        def get_max(node):
            if node is None:
                return 0
            left = max(get_max(node.left), 0)
            right = max(get_max(node.right), 0)
            self.max = max(self.max, node.val + left + right)
            return node.val + max(left, right)
        
        get_max(root)
        return self.max